ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/CSJ.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE LambdaCase     #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Test.Consensus.Genesis.Tests.CSJ (tests) where
+
+import           Control.Monad (replicateM)
+import           Data.Containers.ListUtils (nubOrd)
+import           Data.Functor (($>))
+import           Data.List (nub)
+import           Data.Maybe (mapMaybe)
+import           Ouroboros.Consensus.Block (blockSlot, succWithOrigin)
+import           Ouroboros.Consensus.MiniProtocol.ChainSync.Client
+                     (TraceChainSyncClientEvent (..))
+import           Ouroboros.Consensus.Util.Condense (PaddingDirection (..),
+                     condenseListWithPadding)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import           Test.Consensus.BlockTree (BlockTree (..))
+import           Test.Consensus.Genesis.Setup
+import           Test.Consensus.Genesis.Tests.Uniform (genUniformSchedulePoints)
+import           Test.Consensus.PeerSimulator.Run (SchedulerConfig (..),
+                     defaultSchedulerConfig)
+import           Test.Consensus.PeerSimulator.StateView (StateView (..))
+import           Test.Consensus.PeerSimulator.Trace (TraceEvent (..))
+import           Test.Consensus.PointSchedule
+import           Test.Consensus.PointSchedule.Peers (Peer (..), Peers (..),
+                     mkPeers)
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock (Header, TestBlock)
+import           Test.Util.TestEnv (adjustQuickCheckMaxSize)
+
+tests :: TestTree
+tests =
+  adjustQuickCheckMaxSize (`div` 5) $
+  testGroup
+    "CSJ"
+    [ testProperty "“Happy Path” synchronous scenario only speaks to one peer" $ prop_happyPath True
+    , testProperty "“Happy Path” asynchronous scenario only speaks to one peer" $ prop_happyPath False
+    ]
+
+prop_happyPath :: Bool -> Property
```

Comment:
Could we have the property being tested documented here?

Reply:
Also, I think it needs to be explained what are "synchronous" and "asynchronous" scenarios, which seemed to motivate the boolean parameter.

Reply:
84fce1909fe34888c57f2c153f757d4092b9f65e

Reply:
Thanks much!